text/0015-Service-Reference-Retrieval.md

```diff
@@ -0,0 +1,231 @@
+- Start Date: 2024-10-10
+- CppMicroServices PR: https://github.com/CppMicroServices/CppMicroServices/pull/1044
+
+# Add an free API in that can retrieve a <code>ServiceReference</code> from a <code>std::shared_ptr\<T\></code> given to the client by the framework
+
+## Summary
+
+Currently there is no API in the <code>cppmicroservices</code> namespace that can retrieve a <code>ServiceReference</code> from the <code>std::shared_ptr\<T\></code> itself.
+This document discusses need for such API in the particular use case of declarative services (DS) and its possible design/solution.
+
+## Motivation
+
+Developers using DS rely on DS's SCR (service component runtime) to inject services statically at construction or dynamically using the <code>bind</code> and <code>unbind</code> methods. These <code>std::shared_ptr\<T\></code>s can then be used by the client. However, clients often use properties in the Service to store metadata relevant to that service. Those properties, however, are not available to the <code>std::shared_ptr\<T\></code> itself, only via its <code>ServiceReference</code>.
+
+Given this it seems appropriate to have a new method that the client can use to translate a <code>std::shared_ptr\<T\></code> into its <code>ServiceReference</code>.
+
+## Requirements
+
+This API must:
+
+1. take in a <code>ServiceObject\<T\></code> and return the <code>ServiceReference\<T\></code> for that service where <code>T</code> is any class interface
+2. not expose internal implementation details to the user
+
+## Detailed design
+
+<code>ServiceReferenceFromService</code> API will be declared in the file <i>framework/include/cppmicroservices/ServiceReference.h</i> in the namespace <code><bold>cppmicroservices::</bold></code> as below:
+
+```c++
+
+/**
+* \ingroup MicroServices
+* \ingroup gr_servicereference
+*
+* A method to retrieve a <code>ServiceObject</code>'s original <code>ServiceReference<void></code>
```

Comment:
its confusing to me to reference a `ServiceObject<T>` here. A `ServiceObject<T>` is a cppms framework class used to access service objects registered at the prototype scope. I would recommend using the general term `service object`.